### PR TITLE
overfitting and cross validation

### PR DESCRIPTION
**Pull request by K-sohooli**
Hi @lachlandeer , this is a short article, in which I explained overfitting and cross-validation as a method to detect it.

### CODE REVIEW

**Comment by lachlandeer** [comment]
@K-sohooli 

can you:

- [ ] Add one or two sentences after you do the K-fold to explain how the reader can see what you want them to see
- [ ] Your write "Using cross-validation during model training improves our chances of creating a model that not only fits the training data well but also accurately predicts outcomes for new, unseen data." Can you elaborate in 2-3 additional sentences?

Thanks a lot!

File: content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+description: "In this article, we delve into the concept of overfitting and introduce K-fold cross-validation as a reliable technique for identifying overfitting in models."
+keywords: "machine learning, ML, performance, Python, overfitting, K-fold Crossvalidation, performance, regularization, hyperparameter tuning"
+draft: false
+weight: 20.1
+author: "Kheiry Sohooli"
+authorlink: "https://tilburgsciencehub.com/contributors/kheirysohooli/" 
+---
+
+## Overview
+In the realm of Machine learning, model evaluation is an crucial for assessing their reliability. It helps in detecting overfitting and underfitting after training algorithms on datasets. Cross-validation stands out as the best approach to pinpoint overfitting.
```
Comment (valerievossen) [inline content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md:12]:
```suggestion
In the realm of machine learning, model evaluation is crucial for assessing their reliability. It helps in detecting overfitting and underfitting after training algorithms on datasets. Cross-validation stands out as the best approach to pinpoint overfitting.
```

File: content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+In this short article, you'll become familiar with the primary causes of overfitting and briefly introduce solutions to address it. Additionally, we'll explore K-fold cross-validation, the main method for detecting overfitting.
```
Comment (valerievossen) [inline content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md:13]:
```suggestion
In this short article, you'll become familiar with the primary causes of overfitting and some solutions to address them. Additionally, we'll explore K-fold cross-validation, the main method for detecting overfitting.
```

File: content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+## Overfitting 
+Overfitting occurs when a model performs well on the training data but poorly on the test data. For example, in the image below, the training accuracy is close to 1, while the test accuracy is much lower, less than 0.7. In this case, the model fails to generalize to new samples.
+<div align="center">
```
Comment (valerievossen) [inline content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md:17]: I think there is no image uploaded in this branch yet?

File: content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  <img src="../images/overfitting.png" width="400">
+</div> 
+
+### Why overfitting happen
```
Comment (valerievossen) [inline content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md:21]:
```suggestion
### Why overfitting happens
```

File: content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+Overfitting generally happens when the model is too complex relative to the amount of available data. For example, consider a neural network model with a complex structure trained on a small dataset. Some of the reasons that cause overfitting are as follows: 
+
+1- Complex Model: When the model includes many features, it can become overly complex, detecting noise as well.
+
+2- Small test dataset: If the test dataset is small and doesn't adequately represent all possible input data, it may not provide a reliable measure of the model's performance.
+
+3- Training Data Errors and Outliers: Errors and outliers in the training data introduce noise. If the algorithm interprets this noise as meaningful patterns, it can negatively impact the model's ability to generalize to new data.
```
Comment (valerievossen) [inline content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md:28]:
```suggestion
1. *Complex Model:* When the model includes many features, it can become overly complex, detecting noise as well.

2. *Small test dataset:* If the test dataset is small and doesn't adequately represent all possible input data, it may not provide a reliable measure of the model's performance.

3. *Training Data Errors and Outliers:* Errors and outliers in the training data introduce noise. If the algorithm interprets this noise as meaningful patterns, it can negatively impact the model's ability to generalize to new data.
```

Comment (valerievossen) [reply content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md:28]: If you format the list with "1.", "2." etc., it displays with an indent, which makes it display nicer and better readable on the website! Also making the bullet point argument italicized (with * around the words) makes it clearer! You can do the same for the solutions list below.

File: content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+### Solutions for overfitting
+Several solutions can mitigate  overfitting:
```
Comment (valerievossen) [inline content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md:31]:
```suggestion
Several solutions can mitigate overfitting:
```

File: content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+1- Reducing Model Complexity: Simplify the model by decreasing the number of parameters or attributes. For example, in linear regression, include only the most predictive features instead of all available features. As another example, simplifying the structure of neural network models by reducing the number of nodes and layers in a complex architecture.
```
Comment (valerievossen) [inline content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md:33]:
```suggestion
1- Reducing Model Complexity: Simplify the model by decreasing the number of parameters or attributes. For example, in linear regression, include only the most predictive features instead of all available features. Another example is simplifying the structure of neural network models by reducing the number of nodes and layers in a complex architecture.
```

File: content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+2- Increasing Sample Size: Gathering more real data can help, although it may be costly. Alternatively, augmentation methods can add samples.
+
+3- Noise Reduction: Clean the data by removing outliers and erroneous data to prevent the model from fitting to noisy data.
+
+4- Regularization: Regularization in linear regression is a technique used to prevent overfitting. When a linear model has large coefficients for its features, it tends to overfit the training data. Regularization terms like L1 and L2 works by penalizing large coefficients, which can decrease them or even set them to zero. This process makes the model smoother and helps prevent overfitting.
```
Comment (valerievossen) [inline content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md:39]: - work instead of works?
- What is L1 and L2? Might be useful to explain, or ignore my comment if it is basic ML :)

File: content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+5- Hyperparameter tuning: By finding the best combination of hyperparameters, the model's complexity can be managed. Examples of hyperparameters include the learning rate in gradient descent and the depth of a decision tree. Overall, hyperparameter tuning helps a balance between model complexity and generalization performance, ultimately reducing the risk of overfitting.
```
Comment (valerievossen) [inline content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md:41]: helps create* or helps manage*, maybe?

File: content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+## K-Fold Cross-Validation
+One method to check how well a model works on new data is using train-test-split. Here, we set aside some data to test the model after training it. But a better method is K-fold cross-validation. In this technique, we split the dataset into k subsets or folds. The model is trained k times, each time using a different fold for testing and the rest for training. This gives us k evaluation scores. Finally, we average these scores to get the overall performance. This approach provides a more reliable evaluation compared to just using train-test-split.
```
Comment (valerievossen) [inline content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md:44]:
```suggestion
One method to check how well a model works on new data is using *train-test-split*. Here, we set aside some data to test the model after training it. But a better method is *K-fold cross-validation*. In this technique, we split the dataset into `k` subsets or folds. The model is trained `k` times, each time using a different fold for testing and the rest for training. This gives us `k` evaluation scores. Finally, we average these scores to get the overall performance. This approach provides a more reliable evaluation compared to just using *train-test-split*.
```

File: content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+Using K-fold cross-validation doesn't directly prevent overfitting, but it aids in detecting and addressing it. Let's explore it further by means of an example.
```
Comment (valerievossen) [inline content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md:45]:
```suggestion
Using K-fold cross-validation doesn't directly prevent overfitting, but it aids in detecting and addressing it. Let's explore it further by using an example.
```

File: content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+We'll use the diabetes dataset from Sklearn and train a DecisionTreeRegressor to predict diabetes progression.
```
Comment (valerievossen) [inline content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md:47]:
```suggestion
We'll use the `diabetes` dataset from `Sklearn` and train a `DecisionTreeRegressor` to predict diabetes progression.
```

File: content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+# Calculate Mean Squared Error (MSE)
+mse_train = mean_squared_error(y_train, y_train_pred)
+mse_test = mean_squared_error(y_test, y_test_pred)
+
+print("MSE for Training Data:", mse_train)
+print("MSE for Test Data:", mse_test)
+```
+{{% /codeblock %}}
+
+The result shows that the training MSE is zero, whereas the test MSE is 4933, indicating a severe case of overfitting. We can see this because the MSE in test set is **much** larger than in the training set. When we  implement K-fold cross-validation in next codeblock, it becomes even evident that the model is overfits the training data.
```
Comment (valerievossen) [inline content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md:85]:
```suggestion
The result shows that the training MSE is zero, whereas the test MSE is `4933`, indicating a severe case of overfitting. We can see this because the MSE in the test set is **much** larger than in the training set. When we  implement K-fold cross-validation in the next code block, it becomes even evident that the model overfits the training data.
```

File: content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+# Calculate Mean Squared Error (MSE) for each fold and take the average
+mse_cross_val = -cv_scores.mean()
+print("MSE with 5-fold cross-validation:", mse_cross_val)
+```
+{{% /codeblock %}}
+
+Using cross-validation during model training improves our chances of creating a model that not only fits the training data well but also accurately predicts outcomes for new, unseen data.
+
+ {{% summary %}}
+- Overfitting: Overfitting definition is provided and we discuss what are causes and how to address it.
```
Comment (valerievossen) [inline content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md:101]:
```suggestion
- *Overfitting:* The overfitting definition is provided and we discuss its causes and solutions.
```

File: content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+- K-fold cross-validation: We explore the advantages of K-fold cross-validation compared to the traditional train-test-split method. We also provide an example of its implementation in Python.
```
Comment (valerievossen) [inline content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md:102]:
```suggestion
- Overfitting: Overfitting definition is provided and we discuss what are causes and how to address it.
- *K-fold cross-validation*: We explored the advantages of K-fold cross-validation compared to the traditional train-test-split method. We also provided an example of its implementation in Python.
```

**Review by valerievossen — COMMENTED** [review]
Hi @K-sohooli, 

I had a read over it, and made some suggestions and comments, mostly regarding the formatting and some grammar/spelling updates. Hope it's useful!
